sycl/include/sycl/ext/intel/experimental/esimd/math.hpp

```diff
@@ -1268,59 +1268,59 @@ ESIMD_NODEBUG ESIMD_INLINE
   return esimd_pack_mask(src_0);
 }
 
-/// Count component-wise the total bits set in source operand.
-template <typename T, int N>
-ESIMD_NODEBUG ESIMD_INLINE
-    typename sycl::detail::enable_if_t<std::is_integral<T>::value,
-                                       simd<uint, N>>
-    esimd_cbit(simd<T, N> src0) {
-  return __esimd_cbit(src0.data());
-}
-
-template <typename T>
-ESIMD_NODEBUG ESIMD_INLINE typename sycl::detail::enable_if_t<
-    std::is_integral<T>::value && detail::is_esimd_scalar<T>::value, uint>
-esimd_cbit(T src) {
-  simd<T, 1> Src = src;
-  simd<uint, 1> Result = esimd_cbit(Src);
-  return Result[0];
-}
+#define __ESIMD_DEFINE_BIT_OP(T1, op_hi, op_lo, cond)                          \
```

Comment:
I'm not sure this version is more readable than the previous one.
I understand your desire to remove duplicated code but there are two downsides I see:
1. Users cannot grep `__esimd_cbit`.
2. Hard to read and mentally unwrap all the defines.

This would be perfectly fine in the `detail` namespace, but since this is public API, I'm not so sure whether it is an improvement.

The downside of duplicating code is that it increases the chance of copy-paste bugs, like the one with `operator--`. In this regard, the version that uses macros is better from a developer standpoint.

Reply:
> I'm not so sure whether it is an improvement.

This is not an improvement - this is a fix, as old code did not compile.

These macro start with __ESIMD, so it is not a problem for user space. Besides, I undef it later. Users can grep `cbit`
Wrapping with macros is pretty widely used in SYCL headers to avoid code duplication.

Reply:
Fixed. Had to rebase to resolve conflicts.